lib/src/model/notifications/local_notification_service.dart

```diff
@@ -0,0 +1,156 @@
+    IsolateNameServer.removePortNameMapping('localNotificationServicePort');
+    IsolateNameServer.registerPortWithName(
+      _receivePort.sendPort,
+      'localNotificationServicePort',
+    );
+    _receivePort.listen(_onReceivePortMessage);
+
+    await _notificationPlugin.initialize(
+      InitializationSettings(
+        android: const AndroidInitializationSettings('logo_black'),
+        iOS: DarwinInitializationSettings(
+          requestAlertPermission: false,
+          requestBadgePermission: false,
+          requestSoundPermission: false,
+          notificationCategories: <DarwinNotificationCategory>[
+            ChallengeNotificationDetails.instance.darwinNotificationCategory,
+          ],
+        ),
+      ),
+      onDidReceiveNotificationResponse: _notificationResponse,
+      onDidReceiveBackgroundNotificationResponse:
+          _notificationBackgroundResponse,
+    );
+    _log.info('[Local Notifications] initialized');
+  }
+
+  void _updateLocalisations() {
+    final l10n = _ref.read(l10nProvider);
+    InfoNotificationDetails(l10n.strings);
+    ChallengeNotificationDetails(l10n.strings);
+  }
+
+  Future<int> show(LocalNotification notification) async {
+    final id = currentId++;
+    await _notificationPlugin.show(
+      id,
+      notification.title,
+      notification.body,
+      notification.notificationDetails,
+      payload: notification.payload,
+    );
+    _log.info(
+      '[Local Notifications] Sent notification: ($id | ${notification.title}) ${notification.body} (Payload: ${notification.payload})',
+    );
+
+    // if the notification implements LocalNotificaitonCallback
+    if (notification is LocalNotificationCallback) {
+      _callbacks[id] = (notification as LocalNotificationCallback).callback;
+      _log.info('[Local Notifications] registered callback for id [$id]');
+    }
+
+    return id;
+  }
+
+  Future<void> cancel(int id) async {
+    _log.info('[Local Notifications] canceled notification id: [$id]');
+    return _notificationPlugin.cancel(id);
+  }
+
+  void _onReceivePortMessage(dynamic message) {
+    try {
+      final data = message as Map<String, Object?>;
+      final response = NotificationResponse(
+        notificationResponseType:
+            NotificationResponseType.values[data['index']! as int],
+        id: data['id'] as int?,
+        actionId: data['actionId'] as String?,
+        input: data['input'] as String?,
+        payload: data['payload'] as String?,
+      );
+      _notificationResponse(response);
+    } catch (e) {
+      _log.severe(
+        '[Local Notifications] failed to parse message from background isoalte',
+      );
+    }
+  }
+
+  void _notificationResponse(NotificationResponse response) {
+    final callback = _callbacks[response.id];
+    if (callback != null) callback(response.actionId, response.payload);
+  }
+
+  @pragma('vm:entry-point')
+  static void _notificationBackgroundResponse(NotificationResponse response) {
```

Comment:
Here instead of sending a message, could we not just use the server challenge API to react to user input?

I know we can't have access to riverpod scope here so we're limited. But if the only action is a network request you can instantiate a one time client. Sth like:
```dart
final client = LichessClient(httpClientFactory());
try {
    // do sth
} finally
  client.close()
}
```

Reply:
LichessClient requires a ref to create.
so now a static instance of the LocalNotificationService is required
```dart
  @pragma('vm:entry-point')
  static void _notificationBackgroundResponse(NotificationResponse response) {
    if (response.payload == null) return;
    final instance = LocalNotificationService.instance;
    if (instance == null) return;

    final payload = NotificationPayload.fromJson(
      jsonDecode(response.payload!) as Map<String, Object?>,
    );
    switch (payload.type) {
      case NotificationType.info:
      case NotificationType.challenge:
        final id = payload.data['id'] as String;
        instance._ref
            .read(challengeRepositoryProvider)
            .decline(ChallengeId(id));
    }
  }
```

Reply:
We certainly don't want to do that, I agree on that point.